modules/phase_field/src/actions/GrandPotentialKernelAction.C

```diff
@@ -244,50 +261,133 @@ GrandPotentialKernelAction::act()
     _problem->addKernel("ACGrGrMulti", kernel_name, params);
   } // for (unsigned int i = 0; i < n_etas + n_grs; ++i)
 
-  // Chemical Potentials
-  for (unsigned int i = 0; i < n_w; ++i)
+  if (mass_conservation) // mass conservation kernels with conc and chempot coupling
   {
-    // SusceptibilityTimeDerivative
-    InputParameters params = _factory.getValidParams("SusceptibilityTimeDerivative");
-    params.set<NonlinearVariableName>("variable") = w_names[i];
-    params.set<MaterialPropertyName>("f_name") = chis[i];
-    params.set<std::vector<VariableName>>("coupled_variables") = v0;
-    params.set<bool>("implicit") = implicity;
-    params.set<bool>("use_displaced_mesh") = displaced_mesh;
-    kernel_name = "ChiDt_" + w_names[i];
-    _problem->addKernel("SusceptibilityTimeDerivative", kernel_name, params);
+    std::vector<VariableName> v3;
+    v3.resize(n_w);
+    for (unsigned int i = 0; i < n_w; ++i)
+      v3[i] = w_names[i];
```

Comment:
```suggestion
```

Reply:
i tried implementing your suggestion, but it seems that because w_names is vector<NonlinearVariableName>, I cannot use it to set the kernel parameters that requires vector<VariableName>, so the assignment fails to compile.